ui/store/actions.ts

```diff
@@ -1663,15 +1623,21 @@ export function cancelTxs(
       const cancellations = txIds.map(
         (id) =>
           new Promise<void>((resolve, reject) => {
-            const actionId = generateActionId();
-            callBackgroundMethod('cancelTransaction', [id, actionId], (err) => {
-              if (err) {
-                reject(err);
-                return;
-              }
-
-              resolve();
-            });
```

Comment:
It took me a little while to figure out how the transaction state would correctly be updated now that we are not calling `cancelTransaction` directly. But I think I get it now:
1. `rejectPendingApproval` is called, with the second parameter being the error created by `ethErrors.provider.userRejectedRequest().serialize()`
2. `this.approvalController.reject` is called with an error constructed from its second parameter
3. Steps 1 and 2 will always be happening while we are awaiting `this._requestApproval` in `_requestTransactionApproval` the transaction controller
4. If the user rejects the request, `this._requestApproval` will throw an error
5. The error handler in `_requestTransactionApproval` will check that the error code matches what was set in step 2, and then call `this._cancelTransaction`

Does that sound right?

Reply:
Also, I notice that we are passing `ethErrors.provider.userRejectedRequest().serialize()` here, but in `rejectPendingApproval` in `metamask-controller.js` we assume that the error parameter is not serialized (`new EthereumRpcError(error.code, error.message, error.data)`).

Does deserialization happen automatically by some part of our ui<>background communication code?

Reply:
Exactly right.

The `serialize` method is a bit misleading as it's not serialising to a string, but rather a simple object with only the public properties remaining, including `code`.

https://github.com/MetaMask/rpc-errors/blob/v4.0.3/src/classes.ts#L44